runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/SplittableDoFnOperator.java

```diff
@@ -141,21 +212,182 @@ public void outputWindowedValue(
+    // when we are shut down
+    continuationTimerService.registerEventTimeTimer(
+        TimerInternals.TimerData.of(
+            "continuation",
+            stateNamespace,
+            BoundedWindow.TIMESTAMP_MAX_VALUE,
+            TimeDomain.EVENT_TIME),
+        BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+  }
+
+  /**
+   * Process a restriction. Returns true if the restriction is exhausted, false otherwise.
+   */
+  private boolean processRestriction(
+      ValueState<WindowedValue<InputT>> elementState,
+      ValueState<RestrictionT> restrictionState,
+      WatermarkHoldState holdState,
+      KV<WindowedValue<InputT>, RestrictionT> elementAndRestriction) {
+    final TrackerT tracker = invoker.invokeNewTracker(elementAndRestriction.getValue());
+    SplittableProcessElementInvoker<InputT, OutputT, RestrictionT, TrackerT>.Result result =
+        processElementInvoker.invokeProcessElement(
+            invoker, elementAndRestriction.getKey(), tracker);
+
+    // Save state for resuming.
+    if (result.getResidualRestriction() == null) {
+      // All work for this element/restriction is completed. Clear state and release hold.
+      elementState.clear();
+      restrictionState.clear();
+      holdState.clear();
+      return true;
+    }
+
+    restrictionState.write(result.getResidualRestriction());
+    Instant futureOutputWatermark = result.getFutureOutputWatermark();
+    if (futureOutputWatermark == null) {
+      futureOutputWatermark = elementAndRestriction.getKey().getTimestamp();
+    }
+    holdState.add(futureOutputWatermark);
+
+    return false;
+  }
+
+  @Override
+  public void onEventTime(
+      InternalTimer<Object, TimerInternals.TimerData> flinkTimer) throws Exception {
+    if (flinkTimer.getTimestamp() != BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
+      // it's not our "last-resort" timer
+      return;
+    }
+
+    TimerInternals.TimerData timer = flinkTimer.getNamespace();
+    StateNamespace stateNamespace = timer.getNamespace();
+
+    ValueState<WindowedValue<InputT>> elementState =
+        stateInternals.state(stateNamespace, elementTag);
+    ValueState<RestrictionT> restrictionState =
+        stateInternals.state(stateNamespace, restrictionTag);
+    WatermarkHoldState holdState = stateInternals.state(stateNamespace, watermarkHoldTag);
+
+    boolean restrictionExhausted;
+    do {
+      // This is not the first ProcessElement call for this element/restriction - rather,
+      // this is a timer firing, so we need to fetch the element and restriction from state.
+      elementState.readLater();
+      restrictionState.readLater();
+
+      KV<WindowedValue<InputT>, RestrictionT> elementAndRestriction =
+          KV.of(elementState.read(), restrictionState.read());
+
+      restrictionExhausted =
+          processRestriction(elementState, restrictionState, holdState, elementAndRestriction);
+
+    } while (!restrictionExhausted);
```

Comment:
Then we forever block here, which is not good. However, this is the last chance we have of executing the splittable DoFn. If we get the +Inf watermark, this means that the upstream operators have shut down and that our operator will also be shutdown by Flink if all incoming elements have been processed (Flink will shutdown an operator if it has no more upstream live operators, it will not block shutdown on outstanding timers). 

This is also the main reason why we cannot simply reuse `ProcessFn` as is.